lib/EarlyFunctionAnalysis/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,188 @@
+/// \file ApplyRegisterStateDeductions.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include "revng/ABI/CallingConventionTrait.h"
+#include "revng/EarlyFunctionAnalysis/ApplyRegisterStateDeductions.h"
+#include "revng/Model/Types.h"
+#include "revng/Support/EnumSwitch.h"
+
+/// Finds last relevant (`Yes` or `Dead`) register out of the provided list.
+///
+/// Returns `std::numeric_limits<size_t>::max()` when no registers from
+/// the list were used.
+template<typename AccessorType>
+size_t findLastUsedIndex(llvm::ArrayRef<model::Register::Values> Registers,
+                         const RegisterStateMap &State,
+                         const AccessorType &Accessor) {
+  for (size_t Index = Registers.size(); Index != 0; --Index)
+    if (model::RegisterState::isYesOrDead(Accessor(State.at(Registers[Index]))))
+      return Index;
+
+  return std::numeric_limits<size_t>::max();
+}
+
+/// Limits the passed range to only the elements before the last one used
+/// \sa findLastUsedIndex()
+template<typename AccessorType>
+llvm::ArrayRef<model::Register::Values>
+getRegisterSubrange(llvm::ArrayRef<model::Register::Values> Registers,
+                    const RegisterStateMap &State,
+                    const AccessorType &Accessor) {
+  auto LastUsedIndex = findLastUsedIndex(Registers, State, Accessor);
+  if (LastUsedIndex == std::numeric_limits<size_t>::max())
+    return {};
+
+  return Registers.slice(0, LastUsedIndex);
+}
+
+template<typename CallingConvention, bool EnsureABIConformance = false>
+std::optional<RegisterStateMap>
+applyPositionBasedDeductions(const RegisterStateMap &State) {
```

Comment:
It verifies whether the state is consistent. The idea is to return `std::nullopt` and let the caller know that what they passed in doesn't satisfy the ABI constraints.

Reply:
Yeah, it's verifying compatibility but it's not making any improvement to the `State`.

Reply:
Because there's not exactly much we can do to improve for position based ABIs. Here's an example, if we know that there's an argument is the second GPR, then the only thing we know is that there's either a value (or pointer to it) in the first GPR or a value in the first vector register. And currently there's no way to represent such a relation in the `RegisterStateMap`. Other than using something like `Maybe` on both, but I don't like the approach at all.

Reply:
So, we can only deduce that there is something either in a GPR **or** in a vector register?
That's quite sad.
I feel like we're leaving behind a significant opportunity.
What if say something like "If in the map there's the GPR, we mark it to yes. If there's the vector register, we mark that one to yes. If we have both, we do nothing".

This not great. After all, all the register absent from the map are implicitly `Maybe`, but on the other hand, if a register is in the map it means that function somehow uses it. And 99% of times the vector register will be unused and the regular one will be employed.

What do you think?

Reply:
To be extra specific, it's **xor** (except from a couple of backwards compatibility cases where it's technically both but they have to contain the same value (32-bit float) and the callee is free to use whichever one it chooses. Not sure how applicable it is.

WAIT! It just occurred to me that technically one of those is `Yes` and the other one is `Dead`, isn't it? Should we just mark both with `YesOrDead` and be done with it?

Reply:
Oh, by the way, as of now, I'm not touching the states of any registers that are not initially in the map, should I add additional `YesOrDead`s? Or does it not matter?

Reply:
> To be extra specific, it's xor

Yupyup, I think what my proposal still applies.

> WAIT! It just occurred to me that technically one of those is Yes and the other one is Dead, isn't it? Should we just mark both with YesOrDead and be done with it?

Why do you think it's dead? Keep in mind that if it's `Dead` it's still an argument, but unused. How would we interpret that when going from RFT to CFT?
I'm not persuaded.

> Oh, by the way, as of now, I'm not touching the states of any registers that are not initially in the map, should I add additional YesOrDeads? Or does it not matter?

Yeah, please introduced them.

Reply:
Ok, I see your point, let's go with your initial suggestion of deciding based on what's already accounted for.

Reply:
Extra register introduction is done (https://github.com/revng/revng/pull/98/commits/4aece718133a0ffbb0915b124c03b4737cec4759).

Reply:
What do you think (https://github.com/revng/revng/pull/98/commits/b2e759e91a6e6a7533b037bb8e950d182c5b087e)?

Reply:
Gone.